src/GamePlayer.cpp

```diff
@@ -826,45 +826,50 @@ void GamePlayer::FindWarehouseForAllJobs(const Job job)
 
 Ware* GamePlayer::OrderWare(const GoodType ware, noBaseBuilding* goal)
 {
-    /// Gibt es ein Lagerhaus mit dieser Ware?
-    nobBaseWarehouse* wh = FindWarehouse(*goal, FW::HasMinWares(ware, 1), false, true);
-
-    if(wh)
+    if(emergency)
     {
-        // Prüfe ob Notfallprogramm aktiv
-        if(!emergency)
-            return wh->OrderWare(ware, goal);
-        else
-        {
-            // Wenn Notfallprogramm aktiv nur an Holzfäller und Sägewerke Bretter/Steine liefern
-            if((ware != GD_BOARDS && ware != GD_STONES) || goal->GetBuildingType() == BLD_WOODCUTTER
-               || goal->GetBuildingType() == BLD_SAWMILL)
-                return wh->OrderWare(ware, goal);
-            else
-                return NULL;
-        }
-    } else // no warehouse can deliver the ware -> check all our wares for lost wares that might match the order
+        if(ware == GD_STONES && goal->GetBuildingType() != BLD_SAWMILL)
+            return NULL;
+
+        if(ware == GD_BOARDS && (goal->GetBuildingType() != BLD_WOODCUTTER && goal->GetBuildingType() != BLD_SAWMILL))
+            return NULL;
+    }
+
+    // Look for available ware in a warehouse
+    unsigned int nearestWarehouse = std::numeric_limits<unsigned>::max();
+    nobBaseWarehouse* wh = FindWarehouse(*goal, FW::HasMinWares(ware, 1), false, true, &nearestWarehouse);
+
+    // Look for ware waiting on the road or on the way to storehouse (if addon is enabled)
+    unsigned nearestWare = std::numeric_limits<unsigned>::max();
+    Ware* bestWare = NULL;
+    const bool improvedWareOrder = gwg->GetGGS().isEnabled(AddonId::IMPROVED_WARE_ORDER);
+    if(!wh || improvedWareOrder)
     {
-        unsigned bestLength = std::numeric_limits<unsigned>::max();
-        Ware* bestWare = NULL;
         BOOST_FOREACH(Ware* curWare, ware_list)
         {
-            if(curWare->IsLostWare() && curWare->type == ware)
+            noBaseBuilding* wareGoal = curWare->GetGoal();
```

Comment:
This is an extra mem-load which is not required in many cases (wrong type, lost ware, addon not enabled) and only used once. So don't preload it here.

Reply:
Dont ask me, why i wrote to goal to a variable, since its not used multiple times.